design/build-clusters.md

```diff
@@ -0,0 +1,176 @@
+	1. identify build cluster and queue information for a build and queue jobs in respective queues.
+
+	2. authorize queue worker from build cluster and allow only authorized jobs from queues wrto build cluster.
+
+### Authentication & Authorization
+Initial phase, we will go with JWT + private and public key authorization. Token expiry will be passed as part of signOption. Periodically cycle private+public key and signOption which has the expiry interval, and this will be a manual step which needs to be co-ordinated between Build cluster admin and Screwdriver team.
+
+UI, API to Scheduler service  - will follow the existing JWT authentication and authorization mechanism.
+
+Queue worker to Scheduler service - Build cluster will pass JWT token encrypted with private key and sign option. 
+Queue worker from build cluster need to identify itself and queue he is requesting messages from. 
+
+#### Example: 
+
+1. Queue worker invokes `getJobs/:buildCluster` api 
+
+2. Create JWT token with below information and pass it alongwith `getJobs/:buildCluster` api header.
+
+    1. Payload: { buildCluster: colo1 }
+    2. Secret: privatekey
+    3. SignOption: algorithm, expiry, etc.
+
+3. Scheduler service 
+    1. gets buildCluster name from the getJobs api parameter
+    2. gets respective buildCluster public key based on the buildCluster name from #1 
+    3. Verifies JWT sent by queue worker
+    4. if legit, then decodes the JWT content 
+    5. pulls message from the respective queue
+    6. responds back build info to be processed to queue worker.   
+
+### Yaml 
+
+```yml
+shared:
+    environment:
+    NODE_ENV: test
+    settings:
+        email:
+    addresses: [test@email.com, test2@email.com]
+    statuses: [SUCCESS, FAILURE]
+    annotations:
+        buildCluster: iOS
+jobs:
+    main:
+        requires: [~pr, ~commit]
+        sourcePaths: ["src/app/", "screwdriver.yaml"]
+        image: node:6
+        steps:
+    - init: npm install
+    - test: npm test
+    publish:
+    requires: main
+    image: node:6
+    steps:
+        - publish: npm publish
+    ...
+```   
+ 
+### New table for build cluster details
+Table: `buildClusters`
+
+Columns:
+
+| Name | Type | Nullable | Primary Key | Unique | Description
+| --- | --- | --- | --- | --- | --- |
+| `id` | integer | no | yes | yes | |
+| `name` | text (100) | no | no | yes | |
+| `scmOrganizations` | text(500) | no | no | no | |
+| `scmContext` | text(200) | no | no | no | |
+| `isActive` | boolean | no | no | no | *0-false or 1-true* |
+| `authKey` | text(100) | no | no | yes | *environment variable name of buildCluster publicKey. Note: every buildCluster publicKey will have unique environment variable name* |
+| `signOption` | text(100) | no | no | yes | *environment variable name of buildCluster signOption. Note: every buildCluster publicKey will have unique environment variable name* |
+| `managedBy` | text(50) | no | no | no | cluster managed by *screwdriver or external* |
```

Comment:
this should  be an email address of cluster admin, and we can use this to send out notifications or communications

Reply:
added a new column for communications